ProcessTree.cpp

```diff
@@ -0,0 +1,525 @@
+                _processes[pid] = process;
+            }
+        }
+    }
+
+    closedir(dr);     
+
+    for (auto p : _processes) {
+        auto process = p.second;
+        if (_processes.count(process->_ppid) > 0) {
+            _processes[process->_ppid]->_children.emplace_back(process->_pid);
+        }
+    }
+
+    for (auto p : _processes) {
+        std::shared_ptr<ProcessTreeItem> process, parent;
+        process = p.second;
+        if (_processes.count(process->_ppid) > 0) {
+            parent = _processes[process->_ppid];
+        } else {
+            parent = nullptr;
+        }
+        while (parent) {
+            process->_ancestors.insert(process->_ancestors.begin(), {parent->_pid, parent->_exe});
+            if (_processes.count(parent->_ppid) > 0) {
+                parent = _processes[parent->_ppid];
+            } else {
+                parent = nullptr;
+            }
+        }
+    }
+
+    for (auto p : _processes) {
+        ApplyFlags(p.second);
+    }
+} 
+
+std::string ProcessTree::ReadFirstLine(const std::string& file)
+{
+    std::ifstream f;
+    std::string line;
+
+    try {
+        f.open(file);
+        std::getline(f, line);
+        f.close();
+    } catch (...) {
+    }
+    return line;
+}
+
+std::string ProcessTree::ReadParam(const std::string& file, const std::string& param)
+{
+    std::ifstream f;
+    std::string line;
+    std::string value;
+
+    f.open(file);
+    while (f.good() && f.is_open() && !f.eof()) {
+        std::getline(f, line);
+        if (!line.compare(0, param.size() + 1, param + ":")) {
+            f.close();
+            value = line.substr(param.size() + 1);
+            size_t first = value.find_first_not_of(" \t");
+            if ( first == std::string::npos) {
+                return value;
+            } else {
+                return value.substr(first);
+            }
+        }
+    }
+    f.close();
+    return value;
+}
+
+std::shared_ptr<ProcessTreeItem> ProcessTree::ReadProcEntry(int pid)
+{
+    std::shared_ptr<ProcessTreeItem> process = std::make_shared<ProcessTreeItem>(ProcessTreeSource_procfs, pid);
+    std::string prefix = std::string("/proc/") + std::to_string(pid) + "/";
+
+    // Check if the pid dir exists
+    DIR *dir = opendir(prefix.c_str());
+    if (!dir) {
+        return nullptr;
+    }
+
+    try {
+        std::string uidline = ReadParam(prefix + "status", "Uid");
+        process->_uid = (int)std::stol(uidline.substr(0, uidline.find('\x09')));
+        std::string gidline = ReadParam(prefix + "status", "Gid");
+        process->_gid = (int)std::stol(gidline.substr(0, gidline.find('\x09')));
+        process->_ppid = std::stoi(ReadParam(prefix + "status", "PPid"));
+        char exepath[PATH_MAX];
+        std::string exefpath = prefix + "exe";
+        if (realpath(exefpath.c_str(), exepath)) {
+            process->_exe = std::string(exepath);
+        }
+
+        std::string cmdline = ReadFirstLine(prefix + "cmdline");
```

Comment:
The /proc/<pid>/cmdline file is binary. Each arg is seperated by a NUL. If the first arg contains a '\n' this will fail to read the rest of the args. Also, Linux doesn't have an explicit limit on cmdline size, so it's possible for a process to have a very large (24+MB) cmdline. Please consider using ProcessInfo which limits cmdline to 16K. Also, please consider using ProcessInfo.format_cmdline() which canonicalizes the cmdline.

Reply:
Replaced with ProcessInfo::Open() as suggested.